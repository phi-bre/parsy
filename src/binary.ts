import {ParsyContext, ParsyParser} from '.';

export abstract class BinaryParser extends ParsyParser {
    public left!: ParsyParser;
    public right!: ParsyParser;

    public set(left: ParsyParser, right: ParsyParser): this {
        this.left = left;
        this.right = right;
        return this;
    }
}

export class AndParser extends BinaryParser {
    public parse(context: ParsyContext): ParsyContext | undefined {
        return this.left.parse(context) && this.right.parse(context);
    }
}

export class OrParser extends BinaryParser {
    public parse(context: ParsyContext): ParsyContext | undefined {
        return this.left.parse(context) || this.right.parse(context);
    }
}
