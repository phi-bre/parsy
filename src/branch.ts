import {Token} from './token';
import {indent} from './utils';

export class Branch extends Array<Branch | Token> {

    constructor(public type: string | number | symbol, public parent?: Branch) {
        super();
    }

    toString(level: number = 0): string {
        const space = indent(level);

        let children = '';
        for (const child of this) {
            children += child.toString(level + 1);
        }

        return space
            + this.type.toString()
            + ' [\n'
            + children
            + space
            + ']\n';
    }
}
