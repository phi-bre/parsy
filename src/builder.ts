import {BuilderOptions, Matcher, Node, State, Tokens} from './index';

const matchers = {};

export function builder(options: BuilderOptions) {
    const state = {matchers} as State;

    if (!options.scopes)
        throw 'No scopes specified.';
    if (!options.rules)
        throw 'No rules specified.';
    if (!options.start)
        throw 'No start specified.';

    convert(state, options.terminals, type => state => {
        const token = state.tokens.next;
        if (token.type === type) {
            return state.node.push(token);
        }
        throw 'Unexpected token: ' + token.type.toString() + '. \nExpected: ' + type;
    });

    convert(state, options.rules, type => state => {
        resolve(options.rules[type], state);
    });

    convert(state, options.scopes, type => state => {
        try {
            state.node = new Node(type, state.node);
            resolve(options.scopes[type], state);
            state.node.parent!.push(state.node);
        } finally {
            state.node = state.node.parent!;
        }
    });

    return function (tokens: Tokens) {
        const root = new Node('root');
        state.node = root;
        state.tokens = tokens;
        resolve(options.start, state);

        if (state.tokens.peek)
            throw 'Unexpected token: ' + state.tokens.next.type.toString() + '\nExpected: EOF';

        return root;
    }
}

/**
 * Used internally to convert config types to matchers.
 * @param state
 * @param types
 * @param reducer
 */
function convert(state: State, types, reducer: (type: string) => Matcher) {
    for (const type in types) {
        if (types.hasOwnProperty(type)) {
            if (type in state.matchers)
                throw 'Token type already defined: ' + type;
            state.matchers[type] = reducer(type);
        }
    }
}

/**
 * Used to register new matchers factories.
 * @param types
 * @param label
 * @param matcher
 */
export function register(types: (string | symbol)[], label: string, matcher: Matcher): symbol {
    if (!types?.length)
        throw 'A matcher must have at least one rule.';

    const symbol = Symbol(label);
    matchers[symbol] = matcher;
    return symbol;
}

/**
 * Used to resolve match type references to actual matchers.
 * @param type
 * @param state
 */
export function resolve(type: string | symbol, state: State) {
    if (!state.matchers[type as string])
        throw 'Cannot find reference: ' + type.toString();
    return state.matchers[type as string](state);
}

/**
 * Matches a sequence of types.
 * All the rules provided need to match.
 */
export function sequence(...types: (string | symbol)[]) {
    return register(types, 'sequence', state => {
        for (const type of types) {
            resolve(type, state);
        }
    });
}

/**
 * Searches for the first type that matches in sequential order.
 * At least one of the types has to match.
 */
export function alternation(...types: (string | symbol)[]) {
    return register(types, 'alternation', state => {
        const position = state.tokens.position;
        for (const type of types) try {
            return resolve(type, state);
        } catch {
            state.tokens.reset(position);
        }
        throw 'Unexpected token: ' + state.tokens.peek!.type.toString() + '\nExpected to find: ' + types.join(' | ');
    });
}

/**
 * Matches the supplied sequence but recovers if not all of them match.
 * Makes the sequence optional.
 */
export function optional(...types: (string | symbol)[]) {
    return register(types, 'optional', state => {
        const position = state.tokens.position;
        try {
            for (const type of types) {
                resolve(type, state);
            }
        } catch {
            state.tokens.reset(position);
        }
    });
}

/**
 * Matches and repeats matching the supplied sequence.
 * Recovers if the sequence never matches.
 */
export function star(...rules: (string | symbol)[]) {
    const reference = sequence(...rules);
    return register(rules, 'star', state => {
        while (true) {
            const position = state.tokens.position;
            try {
                resolve(reference, state);
            } catch {
                state.tokens.reset(position);
                break;
            }
        }
    });
}

/**
 * Matches and repeats matching the supplied sequence.
 * Requires the sequence to match at least once.
 */
export function plus(...types: (string | symbol)[]) {
    const reference = star(...types);
    return register(types, 'plus', state => {
        resolve(reference, state);
        // TODO
    });
}
