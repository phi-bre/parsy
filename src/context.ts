import {ParsyToken} from '.';

export class ParsyContext {
    public input: string;
    public index: number;
    public token!: ParsyToken;

    constructor(input: string, index: number = 0) {
        this.input = input;
        this.index = index;
        this.token = new ParsyToken(index);
        this.token.to = index;
    }

    public advance(length: number): this {
        this.index += length;
        return this;
    }

    public open(type?: string): this {
        this.token = new ParsyToken(this.index, this.token, type);
        return this;
    }

    public close(type?: string): this {
        if (!this.token.parent) throw 'Scope mismatch';
        this.token.value = this.input.substring(this.token.from, this.index);
        this.token.to = this.index;
        this.token.parent.children.push(this.token);
        this.token = this.token.parent;
        return this;
    }

    public discard(): this {
        if (!this.token.parent) throw 'Scope mismatch';
        this.token = this.token.parent;
        return this;
    }
}
