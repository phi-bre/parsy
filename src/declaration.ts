import {Node} from './node';
import {Parser} from './parser';

export abstract class Declaration {
    constructor(public type: string | symbol) {
    }

    public abstract parse(parser: Parser): Node | undefined | void;
}
