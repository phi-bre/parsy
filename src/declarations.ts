import {Rule} from './parser';

/**
 * Matches a sequence of rules.
 * All the rules provided need to match.
 */
export function sequence(...rules: (string | Rule)[]): Rule {
    return function (parser) {
        for (const rule of rules) {
            parser.resolve(rule);
        }
    }
}

/**
 * Searches for the first type that matches in sequential order.
 * At least one of the rules has to match.
 */
export function alternation(...rules: (string | Rule)[]): Rule {
    return function (parser) {
        const position = parser.lexer.position;
        for (const rule of rules) {
            try {
                parser.resolve(rule);
                return;
            } catch {
                parser.lexer.reset(position);
            }
        }
        throw `Unexpected token: ${parser.lexer.peek!.type.toString()}. Expected to find: ${rules.join(', ')}`;
    }
}

/**
 * Matches the supplied sequence but recovers if not all of the rules match.
 * Makes the sequence optional.
 */
export function optional(...rules: (string | Rule)[]): Rule {
    const seq = sequence(...rules);
    return function (parser) {
        const position = parser.lexer.position;
        try {
            seq(parser);
        } catch {
            parser.lexer.reset(position);
            return false;
        }
    }
}

/**
 * Matches and repeats matching the supplied sequence.
 * Recovers if the sequence never matches.
 */
export function star(...rules: (string | Rule)[]): Rule {
    const seq = optional(sequence(...rules));
    return function (parser) {
        while (seq(parser) !== false) ;
    }
}

/**
 * Matches and repeats matching the supplied sequence.
 * Requires the sequence to match at least once.
 */
export function plus(...rules: (string | Rule)[]): Rule {
    const seq = sequence(...rules);
    const opt = optional(seq);
    return function (parser) {
        seq(parser);
        while (opt(parser) !== false) ;
    }
}
