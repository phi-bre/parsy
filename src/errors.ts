import {Position, Token} from './';
import {indent} from './utils';

function report(input: string, {index, line, column}: Position) {
    const info = '\nindex: ' + index + ' line: ' + line + ' column: ' + column;
    let from = input.substring(0, index).lastIndexOf('\n');
    let to = input.indexOf('\n', index);
    if (from === -1) from = 0;
    if (to === -1) to = input.length;
    const subject = '\n' + input.substring(from, to);
    const underline = '\n' + indent(index - 1, ' ') + '~~~'; // TODO
    return info + subject + underline;
}

export class TokenError extends Error {
    public severity = 1;

    constructor(input: string, position: Position) {
        super(
            'Unexpected token: "'
            + input.substr(position.index, 1)
            + '"'
            + report(input, position)
        );
    }
}

export class SyntaxError extends Error {
    public severity = 1;

    constructor(input: string, token: Token, expected: string) {
        super(
            'Unexpected token: "'
            + token.value
            + '". expected: '
            + expected
            + report(input, token.position)
        );
    }
}
