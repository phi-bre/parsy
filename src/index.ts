import {Declaration} from './declaration';
import {Lexer} from './lexer';
import {Parser} from './parser';
import {Alternation, Optional, Range, Rule, Scope, Sequence} from './rule';
import {RegexTerminal, StringTerminal, Terminal} from './terminal';

export * from './parser';
export * from './lexer';
export * from './node';
export * from './token';

export interface Position {
    index: number;
    line: number;
    column: number;
}

export interface Declarations {
    [declaration: string]: Declaration;
}

export function parsy(config: { declarations: Declarations, start: Rule, ignore: RegExp }) {
    const rules: Scope[] = [];
    const terminals: Terminal[] = [];
    const declarations: Declarations = {};

    for (const type in config.declarations) {
        const rule = declarations[type] = config.declarations[type];
        rule.type = type;

        if (rule instanceof Rule) {
            rules.push(rule);
        } else if (rule instanceof Terminal) {
            terminals.push(rule);
        }
    }

    for (const rule of rules) {
        rule.resolve(declarations);
        for (const declaration of rule.declarations) {
            declarations[declaration.type as unknown as string] = declaration;
            if (declaration instanceof Terminal) {
                terminals.push(declaration);
            } else if (declaration instanceof Scope) {
                rules.push(declaration);
            }
        }
    }

    config.start.resolve(declarations);

    const ignore = regex(config.ignore);
    terminals.push(ignore);

    return function (input: string) {
        const lexer = new Lexer(terminals, input);
        const parser = new Parser(config.start, lexer);
        return parser.tree;
    }
}


export function string(value: string) {
    return new StringTerminal(Symbol('string'), value);
}

export function regex(value: RegExp) {
    return new RegexTerminal(Symbol('regex'), value);
}

export function token(value: string | RegExp) {
    return typeof value === 'string' ? string(value) : regex(value);
}

export function alias(type: string, rule: Declaration) {
    rule.type = type;
    return rule;
}

export function rule(type: string, rule: Declaration) {
    return new Scope(type, [rule]);
}

/**
 * Matches a sequence of rules.
 * All the rules provided need to match.
 */
export function sequence(...rules: (Declaration | string)[]) {
    return new Sequence(Symbol('sequence'), rules);
}

/**
 * Searches for the first type that matches in sequential order.
 * At least one of the rules has to match.
 */
export function alternation(...rules: (Declaration | string)[]) {
    return new Alternation(Symbol('alternation'), rules);
}

/**
 * Matches the supplied sequence but recovers if not all of the rules match.
 * Makes the sequence optional.
 */
export function optional(...rules: (Declaration | string)[]) {
    return new Optional(Symbol('optional'), rules);
}

/**
 * Matches and repeats matching the supplied sequence.
 * Recovers if the sequence never matches.
 */
export function star(...rules: (Declaration | string)[]) {
    return new Range(Symbol('star'), rules);
}

/**
 * Matches and repeats matching the supplied sequence.
 * Requires the sequence to match at least once.
 */
export function plus(...rules: (Declaration | string)[]) {
    return new Range(Symbol('plus'), rules, 1);
}

/**
 * Matches the specified rules at least 'min' times and repeats for 'max' times.
 */
export function range(min: number, max: number, ...rules: (Declaration | string)[]) {
    return new Range(Symbol('star'), rules, min, max);
}
