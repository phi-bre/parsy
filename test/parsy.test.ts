import {alternation, optional, parsy, sequence, star, token} from '../src';

describe('#parsy', () => {

    const instance = parsy({
        ignore: /[ \n\r]+/,
        start: sequence('value'),
        declarations: {
            left_brace: token('{'),
            right_brace: token('}'),
            left_bracket: token('['),
            right_bracket: token(']'),
            comma: token(','),
            colon: token(':'),
            null: token('null'),
            number: token(/(\d+)?\.?\d+/),
            boolean: token(/true|false/),
            string: token(/"(?:[^"\\]|\\.)*"/),
            value: alternation('null', 'number', 'boolean', 'string', 'object', 'array'),
            property: sequence('string', 'colon', 'value'),
            object: sequence('left_brace', optional('property', star('comma', 'property')), 'right_brace'),
            array: sequence('left_bracket', optional('value', star('comma', 'value')), 'right_bracket'),
        },
    });

    it('should not crash', () => {
        expect(instance('[10, false, {"test": [null]}]').toString()).toMatchSnapshot();
    });
});
