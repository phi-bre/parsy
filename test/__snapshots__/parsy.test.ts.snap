// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parsy should not crash 1`] = `
"root [
  value [
    array [
      left_bracket: [
      value [
        number: 10
      ]
      comma: ,
      value [
        boolean: false
      ]
      comma: ,
      value [
        object [
          left_brace: {
          property [
            string: \\"test\\"
            colon: :
            value [
              array [
                left_bracket: [
                value [
                  null: null
                ]
                right_bracket: ]
              ]
            ]
          ]
          right_brace: }
        ]
      ]
      right_bracket: ]
    ]
  ]
]
"
`;
